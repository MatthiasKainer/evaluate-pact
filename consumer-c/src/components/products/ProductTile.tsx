import React from "react";

import { Product } from "../../types";
import { Heading, Image, Text } from "../common";

interface ProductTileProps {
    readonly product: Product;
}

export function ProductTile(props: ProductTileProps) {
    const {product} = props;
    const firstImage = product.imageUrls[0] ?? "/assets/fallback.png";
    return (
        <div>
            <Image src={firstImage}/>
            <Heading>{product.name}</Heading>
            <Text>{product.description}</Text>
            <Text bold={true}>{product.price}</Text>
        </div>
    );
}