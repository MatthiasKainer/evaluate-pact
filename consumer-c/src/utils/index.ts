export { generateUrl, UrlParams } from "./generateUrl";
export { isEmpty } from "./isEmpty";