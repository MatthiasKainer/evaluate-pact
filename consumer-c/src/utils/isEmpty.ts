export function isEmpty( input: any ): boolean {
    if( input === undefined || input === null || (typeof input === "string" && input === "" ) ) {
        return true;
    }
    if( Array.isArray( input ) && input.length === 0 ) {
        return true;
    }
    return false;
}