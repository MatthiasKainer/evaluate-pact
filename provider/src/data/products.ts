const faker = require("faker")
import { Product } from "../../src/types";

const {commerce, datatype, image } = faker;
faker.seed(5678);

function createImageUrls(): string [] {
    const res: string[] = [];
    for (let i = 0; i < 10; i++) {
        res.push(image.imageUrl());
    }
    return res
}

function createProduct(): Product {
    return {
        id: datatype.uuid(),
        name: commerce.productName(),
        price: parseFloat(commerce.price()) * 10,
        isAvailable: datatype.boolean(),
        imageUrls: createImageUrls()
    }
}

function createProducts(): Product [] {
    const res: Product[] = [];
    for (let i = 0; i < 10; i++) {
        res.push(createProduct());
    }
    return res
}

export const products = createProducts();