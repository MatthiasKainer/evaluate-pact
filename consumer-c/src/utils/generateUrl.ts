import { isEmpty } from "./isEmpty";

export type ValidParameterValue = string | undefined | null | boolean | number;

export interface UrlParams {
    readonly query?: Record<string, ValidParameterValue> | null;
    readonly path?: Record<string, string> | null;
}

function isValidParameterValue(value: ValidParameterValue): value is string | number | boolean {
    return !isEmpty(value);
}

function sanitizeQueryParameters(queryParams: Record<string, ValidParameterValue>): Record<string, string> {
    return Object.entries(queryParams).reduce((result, [key, value]) => {
        if (!isValidParameterValue(value)) {
            return result;
        }
        result[key] = value.toString();
        return result;
    }, {} as Record<string, string>);
}

export function generateUrl(path: string, params: UrlParams = {path: null, query: null}): string {
    const {path: pathParams = null, query: queryParams = null} = params;
    let url = path;
    let queryStr = "";
    if (queryParams !== null) {
        queryStr = new URLSearchParams(sanitizeQueryParameters(queryParams)).toString();
        queryStr = queryStr.length > 0 ? "?" + queryStr : "";
    }
    if (pathParams !== null) {
        url = Object.entries(pathParams).reduce((result, [key, value]) => {
            result = result.split(`{${key}`).join(value);
            return result;
        }, url);
    }
    return url + (queryStr ?? "");
}