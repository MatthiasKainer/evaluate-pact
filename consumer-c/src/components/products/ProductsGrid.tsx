import React from "react";

import { useGetProducts } from "../../hooks/useGetProducts";
import { isEmpty } from "../../utils";
import { Heading } from "../common";
import { ServiceStatus } from "../common/ServiceStatus";
import { ProductTile } from "./ProductTile";

interface ProductsGridProps {
    readonly isAvailable?: boolean;
}
export function ProductsGrid( props: ProductsGridProps) {
    const { isAvailable } = props;
    const options = { params: { query: { isAvailable } } };
    const {isPending, isError, data, error} = useGetProducts( options );
    const hasData = !isEmpty(data);
    const elements = hasData ? data!.map(product => {
        return (
            <li key={product.id}>
                <ProductTile product={product}/>
            </li>
        );
    }) : null;
    return (
        <ServiceStatus isPending={isPending} isError={isError} hasData={hasData} error={error}>
            <Heading>Products</Heading>
            <ul>{elements}</ul>
        </ServiceStatus>
    );
}