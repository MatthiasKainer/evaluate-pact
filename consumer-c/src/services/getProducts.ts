import { Product, ServiceOptions } from "../types";
import { generateUrl } from "../utils";

export function getProducts(options?: ServiceOptions): Promise<Product[]> {
    const { baseUrl = "", params } = options ?? {};
    const url = generateUrl(baseUrl + "/v1/products", params );
    return fetch(url, {method: "GET"}).then(response => response.json());
}