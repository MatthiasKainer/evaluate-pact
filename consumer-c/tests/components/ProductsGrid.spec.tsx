import React from "react";
import { products } from "../../fixtures/getProducts/200";
import { ProductsGrid } from "../../src/components/products/ProductsGrid";
import { render, waitFor } from "../utils";
import { Pact, Matchers } from "@pact-foundation/pact";
import { resolve } from "path";

const { like } = Matchers;

const mockedProvider = new Pact({
    consumer: "e2e Consumer C",
    provider: "e2e Provider",
    cors: true,
    log: resolve(process.cwd(), "logs", "mockserver-integration.log"),
    dir: resolve(process.cwd(), "pacts"),
    logLevel: "warn",
    spec: 2,
})

beforeAll(() => mockedProvider.setup().then(options => {
    console.log(`http://localhost:${options.port}`);
    process.env.API_HOST = `http://localhost:${options.port}`;
}));

afterAll(() => mockedProvider.finalize());

afterEach(() => mockedProvider.verify());

describe("consumer-c/components/products/ProductsGrid", () => {
    describe("integration without filters", () => {
        beforeEach(() => {
            mockedProvider.addInteraction({
                state: "has a collection of products",
                uponReceiving: "a request for fetching all products",
                withRequest: {
                    method: "GET",
                    path: "/v1/products",
                },
                willRespondWith: {
                    status: 200,
                    headers: {
                        "Content-Type": "application/json; charset=utf-8",
                    },
                    body: like(products),
                },
            })
        });

        it("should render ProductsGrid component with all products", async () => {
            const {baseElement, getByRole} = render(<ProductsGrid/>);
            expect(getByRole("heading")).toHaveTextContent("FETCHING");

            await waitFor(() => {
                getByRole("list")
            });

            expect(getByRole("list").children).toHaveLength(10);
            // not very usefull i know just for fiddling around
            expect(baseElement).toMatchSnapshot();
        });
    });

    describe("integration with filters", () => {
        beforeEach(() => {
            mockedProvider.addInteraction({
                state: "has a collection of products that are available",
                uponReceiving: "a request for fetching available products",
                withRequest: {
                    method: "GET",
                    path: "/v1/products",
                    query: {isAvailable: "true"}
                },
                willRespondWith: {
                    status: 200,
                    headers: {
                        "Content-Type": "application/json; charset=utf-8",
                    },
                    body: like(products.filter(p => p.isAvailable)),
                },
            })
        });

        it("should render ProductsGrid component with only available products", async () => {
            const {baseElement, getByRole} = render(<ProductsGrid isAvailable={true}/>);
            expect(getByRole("heading")).toHaveTextContent("FETCHING");

            await waitFor(() => {
                getByRole("list")
            });

            expect(getByRole("list").children).toHaveLength(1);
            // not very usefull i know just for fiddling around
            expect(baseElement).toMatchSnapshot();
        });
    });
});