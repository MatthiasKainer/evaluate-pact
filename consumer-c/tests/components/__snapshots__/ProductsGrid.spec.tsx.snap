// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`consumer-c/components/products/ProductsGrid integration with filters should render ProductsGrid component with only available products 1`] = `
<body>
  <div>
    <h2>
      Products
    </h2>
    <ul>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Intelligent Metal Towels
          </h2>
          <p>
            Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support
          </p>
          <p
            style="font-weight: bold;"
          >
            5.6
          </p>
        </div>
      </li>
    </ul>
  </div>
</body>
`;

exports[`consumer-c/components/products/ProductsGrid integration without filters should render ProductsGrid component with all products 1`] = `
<body>
  <div>
    <h2>
      Products
    </h2>
    <ul>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Small Granite Table
          </h2>
          <p>
            Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support
          </p>
          <p
            style="font-weight: bold;"
          >
            77.3
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Handcrafted Granite Chicken
          </h2>
          <p>
            The Apollotech B340 is an affordable wireless mouse with reliable connectivity, 12 months battery life and modern design
          </p>
          <p
            style="font-weight: bold;"
          >
            32.8
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Rustic Steel Chair
          </h2>
          <p>
            Carbonite web goalkeeper gloves are ergonomically designed to give easy fit
          </p>
          <p
            style="font-weight: bold;"
          >
            91.3
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Awesome Plastic Shoes
          </h2>
          <p>
            New ABC 13 9370, 13.3, 5th Gen CoreA5-8250U, 8GB RAM, 256GB SSD, power UHD Graphics, OS 10 Home, OS Office A & J 2016
          </p>
          <p
            style="font-weight: bold;"
          >
            1.2
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Fantastic Wooden Mouse
          </h2>
          <p>
            The Football Is Good For Training And Recreational Purposes
          </p>
          <p
            style="font-weight: bold;"
          >
            52.9
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Ergonomic Frozen Sausages
          </h2>
          <p>
            The beautiful range of Apple Natural√© that has an exciting mix of natural ingredients. With the Goodness of 100% Natural Ingredients
          </p>
          <p
            style="font-weight: bold;"
          >
            15.7
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Intelligent Metal Towels
          </h2>
          <p>
            Ergonomic executive chair upholstered in bonded black leather and PVC padded seat and back for all-day comfort and support
          </p>
          <p
            style="font-weight: bold;"
          >
            5.6
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Incredible Cotton Fish
          </h2>
          <p>
            Andy shoes are designed to keeping in mind durability as well as trends, the most stylish range of shoes & sandals
          </p>
          <p
            style="font-weight: bold;"
          >
            19.7
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Fantastic Frozen Car
          </h2>
          <p>
            New range of formal shirts are designed keeping you in mind. With fits and styling that will make you stand apart
          </p>
          <p
            style="font-weight: bold;"
          >
            54.6
          </p>
        </div>
      </li>
      <li>
        <div>
          <img
            src="http://placeimg.com/640/480"
          />
          <h2>
            Refined Cotton Fish
          </h2>
          <p>
            Boston's most advanced compression wear technology increases muscle oxygenation, stabilizes active muscles
          </p>
          <p
            style="font-weight: bold;"
          >
            92.5
          </p>
        </div>
      </li>
    </ul>
  </div>
</body>
`;
